x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiUnifiedStreamingProcessor.java

```diff
@@ -0,0 +1,288 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.openai;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.xcontent.ChunkedToXContent;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.inference.results.StreamingUnifiedChatCompletionResults;
+import org.elasticsearch.xpack.inference.common.DelegatingProcessor;
+import org.elasticsearch.xpack.inference.external.response.streaming.ServerSentEvent;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.elasticsearch.xpack.inference.external.response.XContentUtils.moveToFirstToken;
+
+public class OpenAiUnifiedStreamingProcessor extends DelegatingProcessor<Deque<ServerSentEvent>, ChunkedToXContent> {
+    public static final String FUNCTION_FIELD = "function";
+    private static final Logger logger = LogManager.getLogger(OpenAiUnifiedStreamingProcessor.class);
+
+    private static final String CHOICES_FIELD = "choices";
+    private static final String DELTA_FIELD = "delta";
+    private static final String CONTENT_FIELD = "content";
+    private static final String DONE_MESSAGE = "[done]";
+    private static final String REFUSAL_FIELD = "refusal";
+    private static final String TOOL_CALLS_FIELD = "tool_calls";
+    public static final String ROLE_FIELD = "role";
+    public static final String FINISH_REASON_FIELD = "finish_reason";
+    public static final String INDEX_FIELD = "index";
+    public static final String OBJECT_FIELD = "object";
+    public static final String MODEL_FIELD = "model";
+    public static final String ID_FIELD = "id";
+    public static final String CHOICE_FIELD = "choice";
+    public static final String USAGE_FIELD = "usage";
+    public static final String TYPE_FIELD = "type";
+    public static final String NAME_FIELD = "name";
+    public static final String ARGUMENTS_FIELD = "arguments";
+    public static final String COMPLETION_TOKENS_FIELD = "completion_tokens";
+    public static final String PROMPT_TOKENS_FIELD = "prompt_tokens";
+    public static final String TOTAL_TOKENS_FIELD = "total_tokens";
+
+    private final Deque<StreamingUnifiedChatCompletionResults.ChatCompletionChunk> buffer = new LinkedBlockingDeque<>();
+
+    @Override
+    protected void onRequest(long n) {
```

Comment:
It took me a while to grok that `onRequest` is not part of the Flow interface. Maybe call this `upstreamRequest` not to confuse it with the various Flow on* methods

Reply:
Yep sounds good. I still don't quite understand what all that stuff is doing haha. I'll have Pat give an overview when he gets back maybe.